pkg/planner/core/indexmerge_path.go

```diff
@@ -107,6 +116,70 @@ func (ds *DataSource) generateIndexMergePath() error {
 	return nil
 }
 
+func (ds *DataSource) generateNormalIndexPartialPaths4DNF(dnfItems []expression.Expression, usedIndexCount int) (paths []*util.AccessPath, needSelection bool, usedMap []bool, err error) {
+	paths = make([]*util.AccessPath, 0, len(dnfItems))
+	usedMap = make([]bool, len(dnfItems))
+	for offset, item := range dnfItems {
+		cnfItems := expression.SplitCNFItems(item)
+		pushedDownCNFItems := make([]expression.Expression, 0, len(cnfItems))
+		for _, cnfItem := range cnfItems {
+			if expression.CanExprsPushDown(ds.SCtx().GetSessionVars().StmtCtx,
+				[]expression.Expression{cnfItem},
+				ds.SCtx().GetClient(),
+				kv.TiKV,
+			) {
+				pushedDownCNFItems = append(pushedDownCNFItems, cnfItem)
+			} else {
+				needSelection = true
+			}
+		}
+		itemPaths := ds.accessPathsForConds(pushedDownCNFItems, usedIndexCount)
+		if len(itemPaths) == 0 {
+			// for this dnf item, we couldn't generate an index merge partial path.
+			// (1 member of (a)) or (3 member of (b)) or d=1; if one dnf item like d=1 here could walk index path,
+			// the entire index merge is not valid anymore.
+			return nil, false, usedMap, nil
+		}
+		// prune out global indexes.
+		for i := len(itemPaths) - 1; i >= 0; i-- {
+			path := itemPaths[i]
+			if path.Index != nil && path.Index.Global {
+				itemPaths = append(itemPaths[:i], itemPaths[i+1:]...)
+			}
+		}
```

Comment:
Now we can use `slices.DeleteFunc` to make it simpler.

Reply:
cool